core/node/libp2p.go

```diff
@@ -0,0 +1,591 @@
+	opts.Opts = append(opts.Opts, libp2p.DefaultTransports)
+	return
+}
+
+func P2PQUIC(cfg *config.Config) (opts Libp2pOpts, err error) {
+	if cfg.Experimental.QUIC {
+		opts.Opts = append(opts.Opts, libp2p.Transport(libp2pquic.NewTransport))
+	}
+	return
+}
+
+func P2PNoSecurity() (opts Libp2pOpts) {
+	opts.Opts = append(opts.Opts, libp2p.NoSecurity)
+	// TODO: shouldn't this be Errorf to guarantee visibility?
+	log.Warningf(`Your IPFS node has been configured to run WITHOUT ENCRYPTED CONNECTIONS.
+		You will not be able to connect to any nodes configured to use encrypted connections`)
+	return opts
+}
+
+type P2PHostIn struct {
+	fx.In
+
+	Repo          repo.Repo
+	Validator     record.Validator
+	HostOption    HostOption
+	RoutingOption RoutingOption
+	ID            peer.ID
+	Peerstore     peerstore.Peerstore
+
+	Opts [][]libp2p.Option `group:"libp2p"`
+}
+
+type BaseRouting routing.IpfsRouting
+type P2PHostOut struct {
+	fx.Out
+
+	Host    host.Host
+	Routing BaseRouting
+	IpfsDHT *dht.IpfsDHT
+}
+
+// TODO: move some of this into params struct
+func P2PHost(mctx MetricsCtx, lc fx.Lifecycle, params P2PHostIn) (out P2PHostOut, err error) {
+	opts := []libp2p.Option{libp2p.NoListenAddrs}
+	for _, o := range params.Opts {
+		opts = append(opts, o...)
+	}
+
+	ctx, cancel := context.WithCancel(mctx)
+	lc.Append(fx.Hook{
+		OnStop: func(_ context.Context) error {
+			cancel()
+			return nil
+		},
+	})
+
+	opts = append(opts, libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+		r, err := params.RoutingOption(ctx, h, params.Repo.Datastore(), params.Validator)
+		out.Routing = r
+		return r, err
+	}))
+
+	out.Host, err = params.HostOption(ctx, params.ID, params.Peerstore, opts...)
+	if err != nil {
+		return P2PHostOut{}, err
+	}
+
+	// this code is necessary just for tests: mock network constructions
+	// ignore the libp2p constructor options that actually construct the routing!
+	if out.Routing == nil {
+		r, err := params.RoutingOption(ctx, out.Host, params.Repo.Datastore(), params.Validator)
+		if err != nil {
+			return P2PHostOut{}, err
+		}
+		out.Routing = r
+		out.Host = routedhost.Wrap(out.Host, out.Routing)
+	}
+
+	lc.Append(fx.Hook{
+		OnStop: func(ctx context.Context) error {
+			return out.Host.Close()
+		},
+	})
+
+	// TODO: break this up into more DI units
+	// TODO: I'm not a fan of type assertions like this but the
+	// `RoutingOption` system doesn't currently provide access to the
+	// IpfsNode.
+	//
+	// Ideally, we'd do something like:
+	//
+	// 1. Add some fancy method to introspect into tiered routers to extract
+	//    things like the pubsub router or the DHT (complicated, messy,
+	//    probably not worth it).
+	// 2. Pass the IpfsNode into the RoutingOption (would also remove the
+	//    PSRouter case below.
+	// 3. Introduce some kind of service manager? (my personal favorite but
+	//    that requires a fair amount of work).
+	if dht, ok := out.Routing.(*dht.IpfsDHT); ok {
```

Comment:
Can we have different DI units provide different routers and then have a single unit collect them all? (then we can just drop the routing option system)

Reply:
This seems to run into a circular dependency - If we want to construct routing after the host, we'd also need to somehow defer enabling autorelay.

Autorelay construction depends on routing [[1]](https://github.com/libp2p/go-libp2p/blob/master/config/config.go#L187-L205), which in that place can only be provided as `libp2p.Option` to `libp2p.New`. We can't get the full routing there as `PubsubRouter` requires `PubSub`, which requires `Host`..

We could try not passing the `AutoRelay` option to Libp2p constructor, and instead construct it manually here, but it feels hacky. Better ideas?

Reply:
That sucks... we'll fix this when we add fx support to libp2p but I'd like to merge this first.

What if we had _this_ option depend on `IpfsDHT` specifically and then constructed the real router after we construct the host?

Reply:
> What if we had this option depend on IpfsDHT specifically and then constructed the real router after we construct the host?

What about NilRouter? (used when `--routing=none` is set)

I'd say we'll clean this up in the second stage of the refactor (and remove RoutingOption)